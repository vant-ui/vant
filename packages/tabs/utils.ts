import { raf } from '../utils/dom/raf';

export function scrollLeftTo(el: HTMLElement, to: number, duration: number) {
  let count = 0;
  const from = el.scrollLeft;
  const frames = Math.round((duration * 1000) / 16);

  function animate() {
    el.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      raf(animate);
    }
  }

  animate();
}
