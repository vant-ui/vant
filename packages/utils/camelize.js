const camelizeRE = /-(\w)/g;
export default function(str) {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '');
}
