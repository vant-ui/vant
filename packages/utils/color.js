export const RED = '#f44';
export const BLUE = '#1989fa';
export const GREEN = '#4b0';
export const GRAY_DARK = '#969799';
