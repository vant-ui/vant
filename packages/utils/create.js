/**
 * Create a component with common options
 */
import '../locale';
import Icon from '../icon';
import i18n from '../mixins/i18n';
import install from './install';
import Loading from '../loading';

export default function(sfc) {
  sfc.install = sfc.install || install;
  sfc.mixins = sfc.mixins || [];
  sfc.mixins.push(i18n);
  sfc.components = sfc.components || {};
  sfc.components.icon = Icon;
  sfc.components.loading = Loading;

  return sfc;
};
