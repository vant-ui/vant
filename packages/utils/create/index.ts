import { createBEM } from './bem';
import { createComponent } from './component';
import { createI18N } from './i18n';

type CreateNamespaceReturn = [
  ReturnType<typeof createComponent>,
  ReturnType<typeof createBEM>,
  ReturnType<typeof createI18N>
];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = 'van-' + name;
  return [createComponent(name), createBEM(name), createI18N(name)];
}
