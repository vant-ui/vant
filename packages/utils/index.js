import Vue from 'vue';

export { use, useSlots } from './use';

export const isServer = Vue.prototype.$isServer;

export function noop() {}

export function isDef(value) {
  return value !== undefined && value !== null;
}

export function isObj(x) {
  const type = typeof x;
  return x !== null && (type === 'object' || type === 'function');
}

export function get(object, path) {
  const keys = path.split('.');
  let result = object;

  keys.forEach(key => {
    result = isDef(result[key]) ? result[key] : '';
  });

  return result;
}

const camelizeRE = /-(\w)/g;
export function camelize(str) {
  return str.replace(camelizeRE, (_, c) => c.toUpperCase());
}

export function isAndroid() {
  /* istanbul ignore next */
  return isServer ? false : /android/.test(navigator.userAgent.toLowerCase());
}

export function range(num, min, max) {
  return Math.min(Math.max(num, min), max);
}
