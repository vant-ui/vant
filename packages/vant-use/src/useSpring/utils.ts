import { SpringConfig } from './config';

export const now =
  typeof performance !== 'undefined'
    ? performance.now.bind(performance)
    : Date.now.bind(Date);

export const isArray = Array.isArray.bind(Array);

const reusedTuple: [number, number] = [0, 0];
export function stepper(
  secondPerFrame: number,
  x: number,
  v: number,
  destX: number,
  spring: SpringConfig,
): [number, number] {
  // Spring stiffness, in kg / s^2

  // for animations, destX is really spring length (spring at rest). initial
  // position is considered as the stretched/compressed position of a spring
  const Fspring = -spring.tension * (x - destX);

  // Damping, in kg / s
  const Fdamper = -spring.friction * v;

  const a = (Fspring + Fdamper) / (spring.mass || 1);

  const newV = v + a * secondPerFrame;
  const newX = x + newV * secondPerFrame;

  const precision = spring.precision || 0.01;

  if (Math.abs(newV) < precision && Math.abs(newX - destX) < precision) {
    reusedTuple[0] = destX;
    reusedTuple[1] = 0;
    return reusedTuple;
  }

  reusedTuple[0] = newX;
  reusedTuple[1] = newV;

  return reusedTuple;
}
