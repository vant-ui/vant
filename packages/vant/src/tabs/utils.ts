import { raf, cancelRaf } from '@vant/use';
import { ScrollElement, getScrollTop, setScrollTop } from '../utils';

let scrollLeftToRafId: number;
let scrollTopToRafId: number;

export function scrollLeftTo(
  scroller: HTMLElement,
  to: number,
  duration: number
) {
  cancelRaf(scrollLeftToRafId);

  let count = 0;
  const from = scroller.scrollLeft;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);

  function animate() {
    scroller.scrollLeft += (to - from) / frames;

    if (++count < frames) {
      scrollLeftToRafId = raf(animate);
    }
  }

  animate();
}

export function scrollTopTo(
  scroller: ScrollElement,
  to: number,
  duration: number,
  callback: () => void
) {
  cancelRaf(scrollTopToRafId);

  let current = getScrollTop(scroller);
  const isDown = current < to;
  const frames = duration === 0 ? 1 : Math.round((duration * 1000) / 16);
  const step = (to - current) / frames;

  function animate() {
    current += step;

    if ((isDown && current > to) || (!isDown && current < to)) {
      current = to;
    }

    setScrollTop(scroller, current);

    if ((isDown && current < to) || (!isDown && current > to)) {
      scrollTopToRafId = raf(animate);
    } else if (callback) {
      scrollTopToRafId = raf(callback as FrameRequestCallback);
    }
  }

  animate();
}
