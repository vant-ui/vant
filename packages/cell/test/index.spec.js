import Cell from '..';
import { mount } from '../../../test/utils';

test('click event', () => {
  const click = jest.fn();
  const wrapper = mount(Cell, {
    context: {
      on: {
        click
      }
    }
  });

  wrapper.trigger('click');
  expect(click.mock.calls.length).toBeTruthy();
});
