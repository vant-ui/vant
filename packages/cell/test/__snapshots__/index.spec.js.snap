// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cell correctly 1`] = `
<section class="van-doc-demo-section demo-cell" style="background:undefined;">
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">基础用法</h2>
    <div class="van-cell-group van-hairline--top-bottom">
      <div class="van-cell van-hairline">
        <!---->
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <div class="van-cell__value"><span>内容</span></div>
        <!---->
      </div>
      <div class="van-cell van-hairline">
        <!---->
        <div class="van-cell__title"><span>单元格</span>
          <div class="van-cell__label">描述信息</div>
        </div>
        <div class="van-cell__value"><span>内容</span></div>
        <!---->
      </div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">只设置 value</h2>
    <div class="van-cell-group van-hairline--top-bottom">
      <div class="van-cell van-hairline">
        <!---->
        <!---->
        <div class="van-cell__value van-cell__value--alone"><span>内容</span></div>
        <!---->
      </div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">展示图标</h2>
    <div class="van-cell-group van-hairline--top-bottom">
      <div class="van-cell van-hairline">
        <i class="van-icon van-icon-location van-cell__left-icon">
          <!---->
        </i>
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">展示箭头</h2>
    <div class="van-cell-group van-hairline--top-bottom">
      <div class="van-cell van-cell--clickable van-hairline">
        <!---->
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <!---->
        <i class="van-icon van-icon-arrow van-cell__right-icon">
          <!---->
        </i>
      </div>
      <div class="van-cell van-cell--clickable van-hairline">
        <!---->
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <div class="van-cell__value"><span>内容</span></div>
        <i class="van-icon van-icon-arrow van-cell__right-icon">
          <!---->
        </i>
      </div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">高级用法</h2>
    <div class="van-cell-group van-hairline--top-bottom">
      <div class="van-cell van-cell--clickable van-hairline">
        <i class="van-icon van-icon-shop van-cell__left-icon">
          <!---->
        </i>
        <div class="van-cell__title"><span class="van-cell-text">单元格</span> <span class="van-hairline--surround van-tag van-tag--danger">标签</span></div>
        <div class="van-cell__value"><span>内容</span></div>
        <i class="van-icon van-icon-arrow van-cell__right-icon">
          <!---->
        </i>
      </div>
      <div class="van-cell van-cell--clickable van-hairline">
        <i class="van-icon van-icon-location van-cell__left-icon">
          <!---->
        </i>
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <!---->
        <i class="van-icon van-icon-arrow van-cell__right-icon">
          <!---->
        </i>
      </div>
      <div class="van-cell van-hairline">
        <!---->
        <div class="van-cell__title"><span>单元格</span>
          <!---->
        </div>
        <!---->
        <i class="van-cell__right-icon van-icon van-icon-search">
          <!---->
        </i>
      </div>
    </div>
  </section>
</section>
`;
